shared/controlflow/codeql/controlflow/BasicBlock.qll

```diff
@@ -0,0 +1,258 @@
+   * A basic block, that is, a maximal straight-line sequence of control flow nodes
+   * without branches or joins.
+   */
+  private class BasicBlockImpl extends TBasicBlockStart {
+    /** Gets the location of this basic block. */
+    Location getLocation() { result = this.getFirstNode().getLocation() }
+
+    /** Gets an immediate successor of this basic block, if any. */
+    BasicBlock getASuccessor() { result = this.getASuccessor(_) }
+
+    /** Gets an immediate successor of this basic block of a given type, if any. */
+    BasicBlock getASuccessor(SuccessorType t) {
+      result.getFirstNode() = nodeGetASuccessor(this.getLastNode(), t)
+    }
+
+    /** Gets an immediate predecessor of this basic block, if any. */
+    BasicBlock getAPredecessor() { result.getASuccessor(_) = this }
+
+    /** Gets an immediate predecessor of this basic block of a given type, if any. */
+    BasicBlock getAPredecessor(SuccessorType t) { result.getASuccessor(t) = this }
+
+    /** Gets the control flow node at a specific (zero-indexed) position in this basic block. */
+    cached
+    Node getNode(int pos) { bbIndex(this.getFirstNode(), result, pos) }
+
+    /** Gets a control flow node in this basic block. */
+    Node getANode() { result = this.getNode(_) }
+
+    /** Gets the first control flow node in this basic block. */
+    Node getFirstNode() { this = TBasicBlockStart(result) }
+
+    /** Gets the last control flow node in this basic block. */
+    Node getLastNode() { result = this.getNode(this.length() - 1) }
+
+    /** Gets the length of this basic block. */
+    int length() { result = strictcount(this.getANode()) }
+
+    /**
+     * Holds if this basic block immediately dominates basic block `bb`.
+     *
+     * That is, `bb` is an immediate successor of this basic block and all
+     * paths reaching basic block `bb` from some entry point basic block must
+     * go through this basic block.
+     */
+    predicate immediatelyDominates(BasicBlock bb) { bbIDominates(this, bb) }
+
+    /**
+     * Holds if this basic block strictly dominates basic block `bb`.
+     *
+     * That is, all paths reaching `bb` from some entry point basic block must
+     * go through this basic block and this basic block is different from `bb`.
+     */
+    predicate strictlyDominates(BasicBlock bb) { bbIDominates+(this, bb) }
+
+    /**
+     * Holds if this basic block dominates basic block `bb`.
+     *
+     * That is, all paths reaching `bb` from some entry point basic block must
+     * go through this basic block.
+     */
+    predicate dominates(BasicBlock bb) {
+      bb = this or
+      this.strictlyDominates(bb)
+    }
+
+    /**
+     * Holds if `df` is in the dominance frontier of this basic block. That is,
+     * this basic block dominates a predecessor of `df`, but does not dominate
+     * `df` itself.
+     */
+    predicate inDominanceFrontier(BasicBlock df) {
+      this.dominatesPredecessor(df) and
+      not this.strictlyDominates(df)
```

Comment:
This is a very poor dominance frontier algorithm. There exists an algorithm with much better algorithmic running time.
```
/**
 * The dominance frontier relation for basic blocks.
 *
 * This is equivalent to:
 *
 * ```
 *   bbDominates(x, w.getABBPredecessor()) and not bbStrictlyDominates(x, w)
 * ```
 */
predicate dominanceFrontier(BasicBlock x, BasicBlock w) {
  x = w.getABBPredecessor() and not bbIDominates(x, w)
  or
  exists(BasicBlock prev | dominanceFrontier(prev, w) |
    bbIDominates(x, prev) and
    not bbIDominates(x, w)
  )
}
```

Reply:
The above code snippet is from the Java implementation.

Reply:
Nice, all the more reason to make this code shared.

Reply:
I've switched to that implementation and added a reference I found in the Go implementation of the same algorithm.